File: C++_Version/234Palindrome_Linked_List.cpp
//Given a singly linked list, determine if it is a palindrome.

#include <vector>
#include <iostream>
#include <algorithm>
#include <cstdlib>
#include <functional>
using namespace std;


//* Definition for singly-linked list.
struct ListNode 
{
	int val;
	ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
	bool isPalindrome(ListNode* head) 
	{
		if (head == NULL || head->next == NULL)
			return true;

		ListNode* pfast = head, *pslow = head;          //至少有两个数

		while (pfast->next != NULL && pfast->next->next != NULL)   //寻找中点
		{
			pslow = pslow->next;

			pfast = pfast->next->next;
		}

		pfast = head;
		ListNode *ppre = NULL, *pcur = pslow->next, *pnext;
		
		while (pcur!=NULL)                           //把中点以后的部分反向重新排序
		{
			pnext = pcur->next;
			pcur->next = ppre;
			ppre = pcur;
			pcur = pnext;
		}

		while (ppre != NULL)
		{
			if (pfast->val != ppre->val)
				return false;

			pfast = pfast->next;
			ppre = ppre->next;
		}

		return true;
	}
};

int main()
{
	ListNode* headA = new ListNode(0), *ptemp1, *ptemp2 = headA ;

	int a[] = { 0,1,2,6,8,9,10 };

	for (int i = 1; i < 8; ++i)
	{
		ptemp1 = new ListNode(a[i]);
		ptemp2->next = ptemp1;
		ptemp2 = ptemp1;
	}


	Solution sol1;
	headA = sol1.reverseList(headA);

	system("pause");

	return 0;
}

File: Java_Version/234_Palindrome_Linked_List.java
/*
  Given a singly linked list, determine if it is a palindrome.

  Follow up:
        Could you do it in O(n) time and O(1) space?
*/


class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
        val = x;
    }
}

class NodeWrapper {

    ListNode node;
    boolean isOdd;

    NodeWrapper(ListNode node, boolean isOdd) {
        this.node = node;
        this.isOdd = isOdd;
    }
}

class Solution {

    public static void main(String[] args) {

        ListNode head = new ListNode(1);
        head.next = new ListNode(2);
        head.next.next = new ListNode(3);
        head.next.next.next = new ListNode(4);

        NodeWrapper wrapper = new Solution().getMiddleNode(head);

        System.out.println();
    }

    public boolean isPalindrome(ListNode head) {

        if (head == null || head.next == null) {
            return true;
        }
        if (head.next.next == null) {
            return head.val == head.next.val;
        }

        NodeWrapper middleNodeWrapper = getMiddleNode(head);
        ListNode newHead = reverseList(head, middleNodeWrapper.node);

        if (middleNodeWrapper.isOdd) {
            return isTwoListEqual(newHead, middleNodeWrapper.node.next);
        } else {
            return isTwoListEqual(newHead, middleNodeWrapper.node);
        }
    }

    private NodeWrapper getMiddleNode(ListNode head) {

        ListNode slow = head, fast = head;

        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;
        }

        return fast == null ? new NodeWrapper(slow, false) :
                new NodeWrapper(slow, true);

    }

    private ListNode reverseList(ListNode head, ListNode tail) {

        ListNode lastNode = null, currentNode = head;

        while (currentNode.next != tail) {
            ListNode nextNode = currentNode.next;
            currentNode.next = lastNode;
            lastNode = currentNode;
            currentNode = nextNode;
        }

        currentNode.next = lastNode;
        return currentNode;
    }

    private boolean isTwoListEqual(ListNode head1, ListNode head2) {

        while (head1 != null && head2 != null) {

            if (head1.val != head2.val) {
                return false;
            }

            head1 = head1.next;
            head2 = head2.next;
        }

        return head1 == head2;
    }
}
